rfcs/0000-c-unwind-abi.md

```diff
@@ -0,0 +1,288 @@
+[above][motivation]) to change their `extern` annotations to use the new ABI.
+
+Having separate ABIs for `"C"` and `"C unwind"` may make interface design more
+difficult, especially since this RFC [postpones][future-possibilities]
+introducing coercions between function types using different ABIs.
+
+A single ABI that "just works" with C++ (or any other language that may throw
+exceptions) would be simpler to learn and use than two separate ABIs.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+As explained in [this Inside Rust blog post][inside-rust-requirements], we have
+several requirements for any cross-language unwinding design:
+
+* We need to ensure that Rust may in the future adopt an alternate unwinding
+  mechanism; our design must continue working despite such a change.
+* With the `panic=abort` runtime, we must be able to optimize the binary size by removing most code related
+  to unwinding.
+* Changing the panic runtime from `panic=unwind` to `panic=abort` should generally not
+  cause undefined behavior. (Note that the current proposal does not fully
+  satisfy this criterion.)
+* We cannot change the ABI of functions in the `libc` crate, because this would
+  be a breaking change: function pointers of different ABIs have different
+  types.
+
+[inside-rust-requirements]: https://blog.rust-lang.org/inside-rust/2020/02/27/ffi-unwind-design-meeting.html#requirements-for-any-cross-language-unwinding-specification
+
+## Other proposals discussed with the lang team
+[alternatives]: #other-proposals-discussed-with-the-lang-team
+
+Two other potential designs have been discussed in depth; they are
+explained in [this Inside Rust blog post][inside-rust-proposals]. The design in this
+RFC is referred to as "option 2" in that post.
+
+"Option 1" in that blog post only differs from the current proposal in the
+behavior of a forced unwind across a `"C unwind"` boundary under `panic=abort`.
+Under the current proposal, this type of unwind is permitted, allowing
+`longjmp` and `pthread_exit` to behave "normally" with both the `"C"` and the
+`"C unwind"` ABI across all platforms regardless of panic runtime. If there are
+destructors in the unwound frames, this results in undefined behavior. Under
+"option 1", however, all foreign unwinding, forced or unforced, is caught at
+`"C unwind"` boundaries under `panic=abort`, and the process is aborted. This
+gives `longjmp` and `pthread_exit` surprising behavior on some platforms, but
+avoids that cause of undefined behavior in the current proposal.
+
+The other proposal in the blog post, "option 3", is dramatically different. In
+that proposal, foreign exceptions are permitted to cross `extern "C"`
+boundaries, and no new ABI is introduced.
+
+[inside-rust-proposals]: https://blog.rust-lang.org/inside-rust/2020/02/27/ffi-unwind-design-meeting.html#three-specific-proposals
+
+## Reasons for the current proposal
```

Comment:
I was going to say that we should list out the design constraints that led us here. I guess that this section goes in that direction. This was my list:

* Able to fully remove all unwinding machinery from panic=abort builds. 
* Able to change Rust panics to an alternative implementation beyond native system unwinding if desired, by inserting "translation shims" at "C unwind" call sites. This is possible under either implementation, but such shims would be much more widespread with a single ABI.
* Repair the soundness hole when defining safe Rust functions with "C" ABI that panic (by aborting or translating the panic into a foreign exception). All designs did this equally well, either by aborting or by simply propagating the panic naturally.
* Permit a Rust panic to propagate across foreign frames and re-enter Rust code.
* Support longjmp/forced exceptions with "C" ABI, as many common C functions may unwind in this way. This proposal handles this case so long as there are no destructors, which was deemed adequate, as these features are highly platform dependent and not really suitable for widespread use.

Reply:
Hm, several of these are listed in the blog post; perhaps we should link to that? In this section, I focused on differences compared to the other options we discussed in the design-meeting, rather than on our non-negotiable constraints.

Reply:
We could link to the blog post, although I think it's good for the RFC to be reasonably self-contained. Or we could just copy some of that material into a different section. Honestly, I think I would typically put these sorts of constraints into the introduction